File: backend/src/durability.ts
import { existsSync, readdirSync, readFileSync, rmSync, statSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";

/**
 * Survives a FULL app restart (not just the in-process watchdog restart
 * handleFailure already replays from memory) -- if Caroline.exe gets closed
 * or crashes while a turn is still in flight, everything in the ChatSession
 * object is lost, but the user's message may already be sitting in the
 * conversation with no reply. Persisted here so the next process lifetime
 * can notice and proactively finish it instead of leaving it dangling until
 * the user happens to say something else.
 */
export interface PendingTurn {
  text: string;
  attachments: unknown[];
  submittedAtIso: string;
}

// Per tabId (see server.ts's multi-tab ChatSession map) -- a crash mid-turn
// on one tab must resume into that SAME tab on reconnect, not whichever tab
// happens to reconnect first. sanitizeTabId keeps a caller-supplied tabId
// (ultimately from a WS query string) from escaping the workspace dir.
function sanitizeTabId(tabId: string): string {
  return tabId.replace(/[^a-zA-Z0-9_-]/g, "_") || "default";
}

function pendingTurnPath(workspaceDir: string, tabId: string): string {
  return join(workspaceDir, `pending-turn-${sanitizeTabId(tabId)}.json`);
}

export function savePendingTurn(workspaceDir: string, tabId: string, text: string, attachments: unknown[]): void {
  const turn: PendingTurn = { text, attachments, submittedAtIso: new Date().toISOString() };
  try {
    writeFileSync(pendingTurnPath(workspaceDir, tabId), JSON.stringify(turn, null, 2) + "\n", "utf-8");
  } catch (err) {
    // Best-effort -- worst case this specific restart doesn't auto-resume --
    // but per explicit instruction (2026-09-06), that "worst case" must be
    // visible in the log, not just a silently swallowed possibility.
    console.error(`[caroline] savePendingTurn: write failed for tab ${tabId} (ignored):`, err);
  }
}

export function clearPendingTurn(workspaceDir: string, tabId: string): void {
  try {
    rmSync(pendingTurnPath(workspaceDir, tabId), { force: true });
  } catch (err) {
    console.error(`[caroline] clearPendingTurn: rmSync failed for tab ${tabId} (ignored):`, err);
  }
}

/**
 * Read-only -- does NOT delete the file. Root-caused live (2026-08-31): the
 * previous version deleted on read, called once at module load, well before
 * the resume message actually gets injected (that only happens once a
 * WebSocket client connects, see server.ts). During a run of quick
 * kill-and-restart cycles (confirmed live: the external health watchdog
 * killing a backend that hadn't finished starting yet, repeatedly), a
 * process could read-and-delete this file, then itself get killed before a
 * client ever connected to receive the injection -- silently and
 * permanently losing the user's unanswered message, with no trace anywhere.
 *
 * Deletion is no longer this function's job: the eventual resume message is
 * itself submit()'ed like any other turn, which calls savePendingTurn()
 * again (overwriting this file with the resume text) and clearPendingTurn()
 * normally once that turn actually completes -- the same safety net just
 * keeps covering the resume attempt itself if IT also gets interrupted,
 * rather than a one-shot "delete on read" leaving a asingle window where an
 * interruption loses everything with no recovery path at all.
 */
export function peekPendingTurn(workspaceDir: string, tabId: string): PendingTurn | null {
  const path = pendingTurnPath(workspaceDir, tabId);
  if (!existsSync(path)) return null;
  try {
    return JSON.parse(readFileSync(path, "utf-8")) as PendingTurn;
  } catch (err) {
    console.error(`[caroline] peekPendingTurn: read/parse failed for tab ${tabId} (treating as none):`, err);
    return null;
  }
}

// Matches Claude Code's own project-dir slugging: every ':' and path
// separator becomes '-', one-for-one (confirmed against a real project dir
// name, e.g. "C:\Users\x\...\workspace" -> "C--Users-x-...-workspace"). Used
// both here and by compaction.ts to locate a session's .jsonl transcript
// directly on disk.
export function claudeProjectDir(workspaceDir: string): string {
  const encoded = workspaceDir.replace(/[:\\/]/g, "-");
  return join(homedir(), ".claude", "projects", encoded);
}

// --- per-tab Claude session id, for resume: instead of continue: true ------
// continue:true always resumes "the most recent session for this cwd" --
// fine for a single conversation, but with multiple independent tabs
// sharing one workspace/cwd, every tab's continue:true would race to
// resume the SAME most-recent thread. Each tab instead gets resume:<its own
// stored session id>, captured from the SDK's own session_id field (present
// on every message) the first time each tab's query() starts one.
function tabSessionIdPath(workspaceDir: string, tabId: string): string {
  return join(workspaceDir, `tab-session-${sanitizeTabId(tabId)}.json`);
}

export function loadTabSessionId(workspaceDir: string, tabId: string): string | null {
  const path = tabSessionIdPath(workspaceDir, tabId);
  if (!existsSync(path)) return null;
  try {
    const data = JSON.parse(readFileSync(path, "utf-8")) as { sessionId?: string };
    return data.sessionId ?? null;
  } catch (err) {
    console.error(`[caroline] loadTabSessionId: read/parse failed for tab ${tabId} (treating as none):`, err);
    return null;
  }
}

export function saveTabSessionId(workspaceDir: string, tabId: string, sessionId: string): void {
  try {
    writeFileSync(tabSessionIdPath(workspaceDir, tabId), JSON.stringify({ sessionId }, null, 2) + "\n", "utf-8");
  } catch (err) {
    // Best-effort -- worst case this tab starts a fresh conversation next
    // time -- but per explicit instruction (2026-09-06), that must be
    // visible in the log, not just a silently swallowed possibility.
    console.error(`[caroline] saveTabSessionId: write failed for tab ${tabId} (ignored):`, err);
  }
}

/** Removes the stored resume id entirely -- the next runLoop iteration for
 *  this tab omits `resume` from its query() options, so the SDK starts a
 *  genuinely new session instead of resuming anything. Used by compaction's
 *  backoff path (see compaction.ts's backoffPointerNote) when forkSession
 *  itself cannot produce a usable copy of the current session no matter how
 *  many times it's retried. */
export function clearTabSessionId(workspaceDir: string, tabId: string): void {
  try {
    rmSync(tabSessionIdPath(workspaceDir, tabId), { force: true });
  } catch (err) {
    console.error(`[caroline] clearTabSessionId: rmSync failed for tab ${tabId} (ignored):`, err);
  }
}

// --- per-tab continuity-archive pointer ------------------------------------
// Bug fix (2026-09-09): confirmed live -- resetUnrecoverableSession's own
// archive-reference note only ever landed in the ONE turn it was attached
// to (a one-shot user-turn injection, not part of the system prompt). A
// later turn in the same fresh session has no such note in view at all, so
// Caroline had no way to know "this looks like the start of the
// conversation, but it isn't" -- confirmed live as a real incident (she
// flatly denied having just changed a mailbox password, because that whole
// exchange was several turns back in a session she genuinely has no other
// reason to reconsider). Persisted here (not just held in memory) so it
// survives a full app restart too, and read fresh into EVERY query()'s own
// systemPrompt (see server.ts's continuityPointerInstruction usage) for as
// long as it's set -- a real, standing instruction for the whole session's
// lifetime, not a single message that scrolls out of attention.
function tabContinuityArchivePath(workspaceDir: string, tabId: string): string {
  return join(workspaceDir, `tab-continuity-${sanitizeTabId(tabId)}.json`);
}

export function loadTabContinuityArchive(workspaceDir: string, tabId: string): string | null {
  const path = tabContinuityArchivePath(workspaceDir, tabId);
  if (!existsSync(path)) return null;
  try {
    const data = JSON.parse(readFileSync(path, "utf-8")) as { archivePath?: string };
    return data.archivePath ?? null;
  } catch (err) {
    console.error(`[caroline] loadTabContinuityArchive: read/parse failed for tab ${tabId} (treating as none):`, err);
    return null;
  }
}

export function saveTabContinuityArchive(workspaceDir: string, tabId: string, archivePath: string): void {
  try {
    writeFileSync(tabContinuityArchivePath(workspaceDir, tabId), JSON.stringify({ archivePath }, null, 2) + "\n", "utf-8");
  } catch (err) {
    console.error(`[caroline] saveTabContinuityArchive: write failed for tab ${tabId} (ignored):`, err);
  }
}

export function clearTabContinuityArchive(workspaceDir: string, tabId: string): void {
  try {
    rmSync(tabContinuityArchivePath(workspaceDir, tabId), { force: true });
  } catch (err) {
    console.error(`[caroline] clearTabContinuityArchive: rmSync failed for tab ${tabId} (ignored):`, err);
  }
}

// --- per-tab compaction pointer --------------------------------------------
// Same pattern/reasoning as the continuity-archive pointer above, but for
// routine age-based compaction (see compaction.ts) instead of an
// unrecoverable-session error -- see policies.ts's
// compactionPointerInstruction. Naturally overwritten on every subsequent
// compaction; no separate clear function needed.
function tabCompactionNotePath(workspaceDir: string, tabId: string): string {
  return join(workspaceDir, `tab-compaction-${sanitizeTabId(tabId)}.json`);
}

export function loadTabCompactionNote(workspaceDir: string, tabId: string): { parentPath: string | null; compactedAtIso: string | null } {
  const path = tabCompactionNotePath(workspaceDir, tabId);
  if (!existsSync(path)) return { parentPath: null, compactedAtIso: null };
  try {
    const data = JSON.parse(readFileSync(path, "utf-8")) as { parentPath?: string; compactedAtIso?: string };
    return { parentPath: data.parentPath ?? null, compactedAtIso: data.compactedAtIso ?? null };
  } catch (err) {
    console.error(`[caroline] loadTabCompactionNote: read/parse failed for tab ${tabId} (treating as none):`, err);
    return { parentPath: null, compactedAtIso: null };
  }
}

export function saveTabCompactionNote(workspaceDir: string, tabId: string, parentPath: string, compactedAtIso: string): void {
  try {
    writeFileSync(tabCompactionNotePath(workspaceDir, tabId), JSON.stringify({ parentPath, compactedAtIso }, null, 2) + "\n", "utf-8");
  } catch (err) {
    console.error(`[caroline] saveTabCompactionNote: write failed for tab ${tabId} (ignored):`, err);
  }
}

/**
 * One-time migration for users upgrading from pre-multi-tab Caroline: before
 * this, the single conversation was resumed via continue:true (whichever
 * session was most recently active for this cwd), not a stored id -- so the
 * primary tab (server.ts's PRIMARY_TAB_ID) would otherwise start a BLANK
 * conversation on first launch after the upgrade, silently dropping
 * everything discussed before. Best-effort: reads Claude Code's own session
 * transcript directory for this workspace directly off disk (~/.claude/
 * projects/<encoded-cwd>/*.jsonl) and returns the most recently modified
 * one's id, mirroring what continue:true would have picked. Never throws --
 * worst case (directory layout changes, permissions, anything) the caller
 * just falls through to starting fresh, same as any other new tab.
 */
export function findMostRecentClaudeSessionId(workspaceDir: string): string | null {
  try {
    const projectDir = claudeProjectDir(workspaceDir);
    if (!existsSync(projectDir)) return null;
    let best: { id: string; mtimeMs: number } | null = null;
    for (const entry of readdirSync(projectDir)) {
      if (!entry.endsWith(".jsonl")) continue;
      const mtimeMs = statSync(join(projectDir, entry)).mtimeMs;
      if (!best || mtimeMs > best.mtimeMs) best = { id: entry.slice(0, -".jsonl".length), mtimeMs };
    }
    return best?.id ?? null;
  } catch (err) {
    console.error(`[caroline] findMostRecentClaudeSessionId: failed for ${workspaceDir} (falling through to a fresh session):`, err);
    return null;
  }
}

File: backend-py/app/durability.py
"""Ports backend/src/durability.ts -- survives a FULL app restart (not just
the in-process watchdog restart ChatSession's own failure handling replays
from memory): if Caroline's process gets closed or crashes while a turn is
still in flight, everything in the ChatSession object is lost, but the
user's message may already be sitting in the conversation with no reply.
Persisted here so the next process lifetime can notice and proactively
finish it instead of leaving it dangling.

Three per-tab files under workspaceDir: tab-session-<id>.json (resume id),
pending-turn-<id>.json (a not-yet-answered turn, read-only peek -- see
peek_pending_turn's own docstring for why never delete-on-read), and
tab-continuity-<id>.json (the continuity-archive pointer, written only by
ChatSession's own reset-unrecoverable-session path).
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path

from app.logging_setup import log_event


def _sanitize_tab_id(tab_id: str) -> str:
    """Keeps a caller-supplied tabId (ultimately from a WS query string)
    from escaping the workspace dir."""
    sanitized = re.sub(r"[^a-zA-Z0-9_-]", "_", tab_id)
    return sanitized or "default"


def claude_project_dir(workspace_dir: str) -> Path:
    """Matches Claude Code's own project-dir slugging: every ':' and path
    separator becomes '-', one-for-one. Used to locate a session's .jsonl
    transcript directly on disk."""
    encoded = re.sub(r"[:\\/]", "-", workspace_dir)
    return Path.home() / ".claude" / "projects" / encoded


def openai_transcripts_dir(workspace_dir: str) -> Path:
    """workspace/openai-transcripts/ -- the conversation log of tabs answered
    by the OpenAI engine, in the same JSONL shape Claude Code writes, so every
    history reader works on it unchanged (see session_transcript_path)."""
    return Path(workspace_dir) / "openai-transcripts"


def session_transcript_path(workspace_dir: str, session_id: str) -> Path:
    """Where a session's transcript lives: an OpenAI thread's own log if one
    exists under that id, else Claude Code's file. Ids never collide across
    engines (Claude session ids and Codex thread ids are distinct UUIDs)."""
    own = openai_transcripts_dir(workspace_dir) / f"{session_id}.jsonl"
    return own if own.exists() else claude_project_dir(workspace_dir) / f"{session_id}.jsonl"


def dehydrated_dir(workspace_dir: str) -> Path:
    """workspace/dehydrated/ -- where the PreCompact hook copies a
    pre-compaction transcript so earlier context stays recoverable, and
    where main.py's expand_dehydrated_ref serves files back from. Exported
    so that handler can validate a requested path is really inside it."""
    return Path(workspace_dir) / "dehydrated"


# --- pending turn (crash-mid-turn recovery) --------------------------------

@dataclass
class PendingTurn:
    text: str
    attachments: list
    submitted_at_iso: str


def _pending_turn_path(workspace_dir: str, tab_id: str) -> Path:
    return Path(workspace_dir) / f"pending-turn-{_sanitize_tab_id(tab_id)}.json"


def save_pending_turn(workspace_dir: str, tab_id: str, text: str, attachments: list) -> None:
    from datetime import datetime, timezone
    turn = {"text": text, "attachments": attachments, "submittedAtIso": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")}
    try:
        path = _pending_turn_path(workspace_dir, tab_id)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(json.dumps(turn, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    except Exception as exc:
        # Best-effort -- worst case this specific restart doesn't
        # auto-resume, but that must be visible in the log, not silently
        # swallowed.
        log_event("engine", "save_pending_turn_failed", tab_id=tab_id, error=str(exc))


def clear_pending_turn(workspace_dir: str, tab_id: str) -> None:
    try:
        _pending_turn_path(workspace_dir, tab_id).unlink(missing_ok=True)
    except Exception as exc:
        log_event("engine", "clear_pending_turn_failed", tab_id=tab_id, error=str(exc))


def peek_pending_turn(workspace_dir: str, tab_id: str) -> PendingTurn | None:
    """Read-only -- does NOT delete the file. A previous delete-on-read
    version lost the user's unanswered message permanently during a run
    of quick kill-and-restart cycles (a process could read-and-delete,
    then itself get killed before any WS client connected to receive the
    injection). Deletion isn't this function's job: the eventual resume
    message is itself submitted like any other turn, which calls
    save_pending_turn() again (overwriting this file) and
    clear_pending_turn() normally once that turn completes -- the same
    safety net keeps covering the resume attempt itself if IT also gets
    interrupted."""
    path = _pending_turn_path(workspace_dir, tab_id)
    if not path.exists():
        return None
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        return PendingTurn(text=data["text"], attachments=data.get("attachments") or [], submitted_at_iso=data.get("submittedAtIso", ""))
    except Exception as exc:
        log_event("engine", "peek_pending_turn_failed", tab_id=tab_id, error=str(exc))
        return None


# --- per-tab in-flight background operations, for restart recovery -------
# Per explicit instruction (2026-09-15, "Кэролайн регулярно теряет фоновых
# агентов" -> agreed follow-up fix): app/operations.py's OperationRegistry
# is pure in-memory, process-scoped -- a genuine full backend-process
# restart (crash, a forced relaunch, an in-place update) wipes it
# completely, with zero trace, for ANY operation that had already left
# dispatch()'s fast path. pending-turn-<id>.json alone doesn't cover this:
# it only fires when the TRIGGERING turn itself is still unanswered at
# restart time, but the confirmed live incident (an 8-mailbox cleanup the
# model explicitly promised to report back on) had already replied to the
# user and moved on -- the turn was "done" from the SDK's own perspective,
# only the background operation itself was still running. This is a
# SEPARATE small durability file, one per tab, a dict of operation_id ->
# {tool_name, startedAtIso} for every operation currently past the fast
# path -- written by operations.py's dispatch() when an operation goes
# slow, cleared when it actually finishes (see run()'s own completion).
# Anything still in this file when a tab's session first connects in a
# NEW process lifetime is, by construction, stale (a live process's own
# OperationRegistry would already be tracking it) -- main.py's WS handler
# peeks it the same way it already peeks pending-turn and injects a
# recovery nudge, mirroring that exact pattern.


def _pending_operations_path(workspace_dir: str, tab_id: str) -> Path:
    return Path(workspace_dir) / f"pending-operations-{_sanitize_tab_id(tab_id)}.json"


def _load_pending_operations(workspace_dir: str, tab_id: str) -> dict:
    path = _pending_operations_path(workspace_dir, tab_id)
    if not path.exists():
        return {}
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except Exception as exc:
        log_event("engine", "load_pending_operations_failed", tab_id=tab_id, error=str(exc))
        return {}


def save_pending_operation(workspace_dir: str, tab_id: str, operation_id: str, tool_name: str) -> None:
    from datetime import datetime, timezone

    try:
        operations = _load_pending_operations(workspace_dir, tab_id)
        operations[operation_id] = {"toolName": tool_name, "startedAtIso": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")}
        path = _pending_operations_path(workspace_dir, tab_id)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(json.dumps(operations, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    except Exception as exc:
        # Best-effort -- worst case this specific operation isn't covered
        # by restart recovery, but that must be visible in the log, not
        # silently swallowed.
        log_event("engine", "save_pending_operation_failed", tab_id=tab_id, operation_id=operation_id, error=str(exc))


def clear_pending_operation(workspace_dir: str, tab_id: str, operation_id: str) -> None:
    try:
        operations = _load_pending_operations(workspace_dir, tab_id)
        if operation_id not in operations:
            return
        del operations[operation_id]
        path = _pending_operations_path(workspace_dir, tab_id)
        if operations:
            path.write_text(json.dumps(operations, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
        else:
            path.unlink(missing_ok=True)
    except Exception as exc:
        log_event("engine", "clear_pending_operation_failed", tab_id=tab_id, operation_id=operation_id, error=str(exc))


def peek_pending_operations(workspace_dir: str, tab_id: str) -> list[dict]:
    """Read-only -- does NOT delete the file (same reasoning as
    peek_pending_turn's own docstring: deletion isn't this function's
    job). Returns a list of {operation_id, tool_name, started_at_iso}."""
    operations = _load_pending_operations(workspace_dir, tab_id)
    return [
        {"operation_id": op_id, "tool_name": entry.get("toolName", "?"), "started_at_iso": entry.get("startedAtIso", "")}
        for op_id, entry in operations.items()
    ]


# --- per-tab Claude session id, for resume: instead of continue: true -----
# continue:true always resumes "the most recent session for this cwd" --
# fine for a single conversation, but with multiple independent tabs
# sharing one workspace/cwd, every tab's continue:true would race to
# resume the SAME most-recent thread. Each tab instead gets resume:<its
# own stored session id>, captured from the SDK's own session_id field.

def _tab_session_id_path(workspace_dir: str, tab_id: str) -> Path:
    return Path(workspace_dir) / f"tab-session-{_sanitize_tab_id(tab_id)}.json"


# One file per tab holds the resume id of EACH engine that has run in it
# ("sessionId" = the Claude session, "openaiThreadId" = the Codex thread), so
# switching a tab between engines never loses either conversation.
_SESSION_KEY = {"claude": "sessionId", "openai": "openaiThreadId"}


def _read_tab_session_file(workspace_dir: str, tab_id: str) -> dict:
    path = _tab_session_id_path(workspace_dir, tab_id)
    if not path.exists():
        return {}
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        return data if isinstance(data, dict) else {}
    except Exception as exc:
        log_event("engine", "load_tab_session_id_failed", tab_id=tab_id, error=str(exc))
        return {}


def load_tab_session_id(workspace_dir: str, tab_id: str, engine: str = "claude") -> str | None:
    return _read_tab_session_file(workspace_dir, tab_id).get(_SESSION_KEY[engine])


def save_tab_session_id(workspace_dir: str, tab_id: str, session_id: str, engine: str = "claude") -> None:
    try:
        path = _tab_session_id_path(workspace_dir, tab_id)
        path.parent.mkdir(parents=True, exist_ok=True)
        data = _read_tab_session_file(workspace_dir, tab_id)
        data[_SESSION_KEY[engine]] = session_id
        path.write_text(json.dumps(data, indent=2) + "\n", encoding="utf-8")
    except Exception as exc:
        log_event("engine", "save_tab_session_id_failed", tab_id=tab_id, error=str(exc))


def clear_tab_session_id(workspace_dir: str, tab_id: str, engine: str | None = "claude") -> None:
    """The next runLoop iteration for this tab omits `resume` from its
    query() options, so the SDK starts a genuinely new session instead of
    resuming anything. Used by compaction's backoff path and by
    reset-unrecoverable-session. engine=None clears every engine's id (a full
    tab wipe)."""
    try:
        path = _tab_session_id_path(workspace_dir, tab_id)
        if engine is None:
            path.unlink(missing_ok=True)
            return
        data = _read_tab_session_file(workspace_dir, tab_id)
        if data.pop(_SESSION_KEY[engine], None) is None:
            return
        if data:
            path.write_text(json.dumps(data, indent=2) + "\n", encoding="utf-8")
        else:
            path.unlink(missing_ok=True)
    except Exception as exc:
        log_event("engine", "clear_tab_session_id_failed", tab_id=tab_id, error=str(exc))


# --- per-tab continuity-archive pointer ------------------------------------
# Read fresh into EVERY query()'s own systemPrompt for as long as it's set
# -- a real, standing instruction for the whole session's lifetime, not a
# single message that scrolls out of attention (a one-shot version of this
# caused a confirmed-live incident: the model flatly denied having just
# done something from a few turns back in a fresh session it had no other
# reason to reconsider).

def _tab_continuity_archive_path(workspace_dir: str, tab_id: str) -> Path:
    return Path(workspace_dir) / f"tab-continuity-{_sanitize_tab_id(tab_id)}.json"


def load_tab_continuity_archive(workspace_dir: str, tab_id: str) -> str | None:
    path = _tab_continuity_archive_path(workspace_dir, tab_id)
    if not path.exists():
        return None
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        archive_path = data.get("archivePath")
        # A pointer to a file that no longer exists (an archive pruned as
        # redundant, deleted by hand) is worse than no pointer: the system
        # prompt would send the model to read something that isn't there.
        if archive_path and not Path(archive_path).exists():
            log_event("engine", "tab_continuity_archive_missing", tab_id=tab_id, archive_path=archive_path)
            return None
        return archive_path
    except Exception as exc:
        log_event("engine", "load_tab_continuity_archive_failed", tab_id=tab_id, error=str(exc))
        return None


def save_tab_continuity_archive(workspace_dir: str, tab_id: str, archive_path: str) -> None:
    try:
        path = _tab_continuity_archive_path(workspace_dir, tab_id)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(json.dumps({"archivePath": archive_path}, indent=2) + "\n", encoding="utf-8")
    except Exception as exc:
        log_event("engine", "save_tab_continuity_archive_failed", tab_id=tab_id, error=str(exc))


def clear_tab_continuity_archive(workspace_dir: str, tab_id: str) -> None:
    try:
        _tab_continuity_archive_path(workspace_dir, tab_id).unlink(missing_ok=True)
    except Exception as exc:
        log_event("engine", "clear_tab_continuity_archive_failed", tab_id=tab_id, error=str(exc))


# --- per-tab chat mode (claude vs sw), user-facing Settings toggle ---------
# Per explicit instruction (2026-09-14): each tab independently picks which
# backend answers its real turns -- "claude" (the full Claude Agent SDK,
# submit()) or "sw" (the small-model/Camerlengo path billed through the
# SquirrelWisdom PIA wallet, run_small_model_turn()). Defaults to "claude"
# for every tab; see subscription_mode.py's chat_mode_eligible() for the
# gating (both a Claude subscription AND a paid SW balance) that decides
# whether "sw" can be SET at all -- this module only persists whatever was
# already validated, same file-per-tab shape as tab-session-<id>.json etc.

def _chat_mode_path(workspace_dir: str, tab_id: str) -> Path:
    return Path(workspace_dir) / f"chat-mode-{_sanitize_tab_id(tab_id)}.json"


def load_chat_mode(workspace_dir: str, tab_id: str) -> str:
    path = _chat_mode_path(workspace_dir, tab_id)
    if not path.exists():
        return "claude"
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        mode = data.get("mode")
        return mode if mode in ("claude", "sw", "openai") else "claude"
    except Exception as exc:
        log_event("engine", "load_chat_mode_failed", tab_id=tab_id, error=str(exc))
        return "claude"


def save_chat_mode(workspace_dir: str, tab_id: str, mode: str) -> None:
    try:
        path = _chat_mode_path(workspace_dir, tab_id)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(json.dumps({"mode": mode}, indent=2) + "\n", encoding="utf-8")
    except Exception as exc:
        log_event("engine", "save_chat_mode_failed", tab_id=tab_id, error=str(exc))


def find_most_recent_claude_session_id(workspace_dir: str) -> str | None:
    """One-time migration for users upgrading from pre-multi-tab Caroline:
    reads Claude Code's own session transcript directory for this
    workspace directly off disk and returns the most recently modified
    one's id, mirroring what continue:true would have picked. Never
    raises -- worst case the caller just falls through to starting fresh."""
    try:
        project_dir = claude_project_dir(workspace_dir)
        if not project_dir.exists():
            return None
        best: tuple[str, float] | None = None
        for entry in project_dir.iterdir():
            if entry.suffix != ".jsonl":
                continue
            mtime = entry.stat().st_mtime
            if best is None or mtime > best[1]:
                best = (entry.stem, mtime)
        return best[0] if best else None
    except Exception as exc:
        log_event("engine", "find_most_recent_claude_session_id_failed", workspace_dir=workspace_dir, error=str(exc))
        return None
